tests/test_orient_symmetric.py

```diff
@@ -54,19 +224,113 @@ def testEstimateThirdRows(self):
 
         # Estimate third rows from outer products.
         # Due to factorization of V, these might be negated third rows.
-        vis = self.cl_class._estimate_third_rows(vijs, viis)
+        vis = self.cl_classes[3]._estimate_third_rows(vijs, viis)
 
         # Check if all-close up to difference of sign
         ground_truth = np.sign(gt_vis[0, 0]) * gt_vis
         estimate = np.sign(vis[0, 0]) * vis
         self.assertTrue(np.allclose(ground_truth, estimate))
 
+    @parameterized.expand([(order,), (order + 1,)])
+    def testSelfCommonLines(self, order):
+        n_img = self.n_img
+        n_theta = self.n_theta
+        src = self.srcs[order]
+        cl_symm = self.cl_classes[order]
+
+        # Initialize common-lines class and compute self-common-lines matrix.
+        scl, _, _ = cl_symm._self_clmatrix_c3_c4()
+
+        # Compute ground truth self-common-lines matrix.
+        rots = src.rots
+        scl_gt = self.buildSelfCommonLinesMatrix(rots, order)
+
+        # Get angle difference between scl_gt and scl.
+        scl_diff1 = scl_gt - scl
+        scl_diff2 = scl_gt - np.flip(scl, 1)  # Order of indices might be switched.
+        scl_diff1_angle = scl_diff1 * 2 * pi / n_theta
+        scl_diff2_angle = scl_diff2 * 2 * pi / n_theta
+
+        # cosine is invariant to 2pi, and abs is invariant to +-pi due to J-conjugation.
+        # We take the mean deviation wrt to the two lines in each image.
+        scl_diff1_angle_mean = np.mean(np.arccos(abs(np.cos(scl_diff1_angle))), axis=1)
+        scl_diff2_angle_mean = np.mean(np.arccos(abs(np.cos(scl_diff2_angle))), axis=1)
+
+        scl_diff_angle_mean = np.vstack((scl_diff1_angle_mean, scl_diff2_angle_mean))
+        scl_idx = np.argmin(scl_diff_angle_mean, axis=0)
+        min_mean_angle_diff = scl_idx.choose(scl_diff_angle_mean)
+
+        # Assert scl detection rate is greater than 90% with 1 degree tolerance for order=3,
```

Comment:
Why image sizes of 64×64, I'm not sure it makes sense to look at accuracies of one or three degrees (I don't think rays can be identifies with this accuracy at that resolution). This could be why the detection rate is so low. If there's no noise, should we miss *any* self-common lines apart from discretization (i.e., resolution) errors?

Reply:
Ok, I wasn't sure about what level of accuracy to expect. In the old Python code a 5 degree tolerance is used. At 5 degrees, we find 100% of the self-common lines. I think I need to better understand the effect of `resolution`, `n_rad`, and `n_theta` on the resulting accuracy.

I've changed the angle tolerance to 5 degrees for now. Maybe we can discuss this next meeting.

Reply:
Ok sounds good. You can get some basic intuition by setting up a L×L grid and looking at the angle formed by the points (L // 2, 0) and (L // 2, 1). Anything smaller than that is going to be hard to resolve. Can run some experiments with the CL matrix code to check.